man/man3/Cluster.3
.TH "Cluster" 3 "Dc Mai 20 2020" "Version v6.3 19/05/2020" "Creació d'un arbre filogenètic. Xavier Coll Ribas" \" -*- nroff -*-
.ad l
.nh
.SH NAME
Cluster \- Representa un clúster\&.  

.SH SYNOPSIS
.br
.PP
.SS "Mètodes públics"

.in +1c
.ti -1c
.RI "\fBCluster\fP (const string &id)"
.br
.RI "Crea un clúster a partir d'una espècie\&. "
.ti -1c
.RI "\fBCluster\fP (const \fBCluster\fP clus1, const \fBCluster\fP Clus2, const double &dist)"
.br
.RI "Crea un clúster a partir de la fusió de dos altres\&. "
.ti -1c
.RI "void \fBescriure\fP () const"
.br
.RI "Escriu el clúster que entra pel paràmetre implícit\&. "
.in -1c
.SS "Mètodes Privats Estàtics"

.in +1c
.ti -1c
.RI "static void \fBinm_escriure\fP (const BinTree< pair< string, double >> &a)"
.br
.RI "Funció recursiva cridada per la funcio \fBescriure()\fP per tal de poder fer la escriptura recursiva d'un clúster\&. "
.in -1c
.SS "Atributs Privats"

.in +1c
.ti -1c
.RI "BinTree< pair< string, double > > \fBcluster\fP"
.br
.RI "Estructura de dades d'un clúster\&. 
.br
 "
.in -1c
.SH "Descripció Detallada"
.PP 
Representa un clúster\&. 

Ofereix les opcions d'imprimir l'arbre del clúster amb les distancies respecte als seus fills\&. També l'opció de crear un clúster a partir de la fusió de dos altres\&.
.PP
\fBInvariant\fP
.RS 4
Els clústers contenen una única espècie o bé estàn formats per altres clústers\&. Així mateix, un clúster és assimilable a un arbre binari on té espècies a les fulles\&. 
.RE
.PP

.PP
Definició a la línia 29 del fitxer Cluster\&.hh\&.
.SH "Documentació del Constructor i el Destructor"
.PP 
.SS "Cluster::Cluster (const string & id)"

.PP
Crea un clúster a partir d'una espècie\&. 
.PP
\fBPrecondició\fP
.RS 4
Cert\&. 
.RE
.PP
\fBPostcondició\fP
.RS 4
El resultat és un clúster de la espècie demanada\&. 
.RE
.PP
\fBParàmetres\fP
.RS 4
\fIString\&.\fP Un identificador del clúster\&. 
.RE
.PP

.PP
Definició a la línia 15 del fitxer Cluster\&.cc\&.
.PP
.nf
15                                 {
16     //Es crea un pair on hi ha l'identificador del clúster i
17     //la distancia respecte els seus fills\&. Al ser un clúster sense fills, la distancia es 0\&.
18     pair<string, double> clust;
19     clust\&.first = id;
20     clust\&.second = 0;
21     //Es crea l'arbol binari de clúster\&.
22     cluster = BinTree<pair<string, double>> (clust);
23 }
.fi
.SS "Cluster::Cluster (const \fBCluster\fP clus1, const \fBCluster\fP Clus2, const double & dist)"

.PP
Crea un clúster a partir de la fusió de dos altres\&. 
.PP
\fBPrecondició\fP
.RS 4
Els dos clústers existeixen i no són buits\&. 
.RE
.PP
\fBPostcondició\fP
.RS 4
El resultat és un clúster amb la fusió dels dos altres\&. 
.RE
.PP
\fBParàmetres\fP
.RS 4
\fICluster\&.\fP Dos objectes de tipus clúster\&. 
.RE
.PP

.PP
Definició a la línia 25 del fitxer Cluster\&.cc\&.
.PP
.nf
25                                                                              {
26     //Es fusionen dos clústers en un\&. El identificador del pair és el
27     //resultat de fusionar els dos id's\&.
28     //El double dist és la distancia entre els dos clústers\&.
29     pair<string, double> clust;
30     clust\&.first = clus1\&.cluster\&.value()\&.first + clus2\&.cluster\&.value()\&.first;
31     clust\&.second = dist/2; //dist/2 ja que la formula ho demana així\&.
32     //Es crea un arbre amb node principal el pair acabat de crear, node esquerra el 
33     //arbre del primer clúster i node dret el arbre del segon clúster\&.
34     cluster = BinTree<pair<string,double>> (clust,clus1\&.cluster,clus2\&.cluster);
35 }
.fi
.SH "Documentació de les Funcions Membre"
.PP 
.SS "void Cluster::inm_escriure (const BinTree< pair< string, double >> & a)\fC [static]\fP, \fC [private]\fP"

.PP
Funció recursiva cridada per la funcio \fBescriure()\fP per tal de poder fer la escriptura recursiva d'un clúster\&. 
.PP
\fBPrecondició\fP
.RS 4
Cert\&. 
.RE
.PP
\fBPostcondició\fP
.RS 4
Escriu a en preordre tal com indicat a la pràctica\&. 
.RE
.PP
\fBParàmetres\fP
.RS 4
\fIBinTree\&.\fP Arbre binàri del clúster\&. 
.RE
.PP

.PP
Definició a la línia 53 del fitxer Cluster\&.cc\&.
.PP
.nf
53                                                                  {
54     //Escriptura del arbre binari\&. Si aquest té els nodes 
55     //dret i esquerre buit, es a dir, són fulles de l'arbre,
56     //no escriu la distancia respecte als fills\&.
57       if (not a\&.empty()) {
58         if (not a\&.left()\&.empty() and not a\&.right()\&.empty()) {
59     cout <<"["<< "(" << a\&.value()\&.first << ", " << a\&.value()\&.second<< ") ";
60         }
61         else {
62             cout <<"["<< a\&.value()\&.first;
63         }
64     inm_escriure(a\&.left());
65     inm_escriure(a\&.right());
66         cout << "]";
67   }
68 }
.fi
.SS "void Cluster::escriure () const"

.PP
Escriu el clúster que entra pel paràmetre implícit\&. 
.PP
\fBPrecondició\fP
.RS 4
Cert\&. 
.RE
.PP
\fBPostcondició\fP
.RS 4
S'ha escrit el clúster\&. 
.RE
.PP

.PP
Definició a la línia 45 del fitxer Cluster\&.cc\&.
.PP
.nf
45                              {
46     //Entra el arbre del paràmetre implícit a la funció recursiva\&.
47     inm_escriure(Cluster::cluster);
48     cout << endl;
49 }
.fi
.SH "Documentació de les Dades Membre"
.PP 
.SS "BinTree< pair<string, double> > Cluster::cluster\fC [private]\fP"

.PP
Estructura de dades d'un clúster\&. 
.br
 
.PP
Definició a la línia 35 del fitxer Cluster\&.hh\&.

.SH "Autor"
.PP 
Generat automàticament per Doxygen per a Creació d'un arbre filogenètic\&. Xavier Coll Ribas a partir del codi font\&.


Cluster.cc
/** @file Cluster.cc
 * 	@brief Còdig de la classe Cluster.
 */

	//invariant Els clústers contenen una única espècie o bé estàn formats
	//per altres clústers. Així mateix, un clúster és assimilable a un
	//arbre binari on té espècies a les fulles. 


#include "Cluster.hh"


// Constructores

Cluster::Cluster(const string& id){
    //Es crea un pair on hi ha l'identificador del clúster i
    //la distancia respecte els seus fills. Al ser un clúster sense fills, la distancia es 0.
    pair<string, double> clust;
    clust.first = id;
    clust.second = 0;
    //Es crea l'arbol binari de clúster.
    cluster = BinTree<pair<string, double>> (clust);
}

Cluster::Cluster(const Cluster clus1, const Cluster clus2, const double& dist) {
    //Es fusionen dos clústers en un. El identificador del pair és el
    //resultat de fusionar els dos id's.
    //El double dist és la distancia entre els dos clústers.
    pair<string, double> clust;
    clust.first = clus1.cluster.value().first + clus2.cluster.value().first;
    clust.second = dist/2; //dist/2 ja que la formula ho demana així.
    //Es crea un arbre amb node principal el pair acabat de crear, node esquerra el 
    //arbre del primer clúster i node dret el arbre del segon clúster.
    cluster = BinTree<pair<string,double>> (clust,clus1.cluster,clus2.cluster);
}


// Modificadores

// Consultores

//Lectura i escriptura


void Cluster::escriure() const {
    //Entra el arbre del paràmetre implícit a la funció recursiva.
    inm_escriure(Cluster::cluster);
    cout << endl;
}

// Mètodes privats

void Cluster::inm_escriure(const BinTree<pair<string,double>>& a)  {
    //Escriptura del arbre binari. Si aquest té els nodes 
    //dret i esquerre buit, es a dir, són fulles de l'arbre,
    //no escriu la distancia respecte als fills.
    	if (not a.empty()) {
        if (not a.left().empty() and not a.right().empty()) {
		cout <<"["<< "(" << a.value().first << ", " << a.value().second<< ") ";
        }
        else {
            cout <<"["<< a.value().first;
        }
		inm_escriure(a.left());
		inm_escriure(a.right());
        cout << "]";
	}
}




